### PR TITLE
dynamically support different base token via init

### LINKED ISSUE #708: Implement the feature to support different gas tokens
Currently we only suppport native EOS as gas tokens. Will make it able to support a different token as gas token.

### PR DESCRIPTION
**Pull request by taokayan**
Resolves #708 

- add pararmeter `token_contract` to init action
- make `ingress_bridge_fee` as required field in init().

```
./cleos -u $url --abi-file $evm_acc:./evm_runtime.abi push action $evm_acc init "{\"chainid\":$msig_chain_id,\"fee_params\":{\"gas_price\":150000000000,\"miner_cut\":10000,\"ingress_bridge_fee\":\"0.00005000 BTC\"},\"token_contract\":\"btcbtcbtcbtc\"}" -p $evm_acc@active -s -d -j -x ${trx_exp_sec} > evminit.json
```

### CODE REVIEW

File: include/evm_runtime/tables.hpp [hunk 1]
```diff
@@ -149,6 +157,11 @@ struct [[eosio::table("inevm")]] [[eosio::contract("evm_contract")]] balance_wit
     }
 
     balance_with_dust& operator-=(const intx::uint256& amount) {
+
+        check(balance.symbol != eosio::symbol(), "symbol can't be empty in balance_with_dust");
+        intx::uint256 minimum_natively_representable = pow10_const(evm_precision - balance.symbol.precision());
+        uint64_t min_asset = minimum_natively_representable[0];
+
```
Comment (elmato) [inline include/evm_runtime/tables.hpp:164]: Is there any value in caching the `minimum_natively_representable` value in the config table also to not have to calculate all the time?

Comment (taokayan) [reply include/evm_runtime/tables.hpp:164]: please review again. thanks!

File: include/evm_runtime/types.hpp [hunk 2]
```diff
@@ -18,18 +16,34 @@ namespace evm_runtime {
...
-   constexpr eosio::name token_account(eosio::name(TOKEN_ACCOUNT_NAME));
-   constexpr eosio::symbol token_symbol("EOS", 4u);
-   constexpr intx::uint256 minimum_natively_representable = intx::exp(10_u256, intx::uint256(evm_precision - token_symbol.precision()));
-   static_assert(evm_precision - token_symbol.precision() <= 14, "dust math may overflow a uint64_t");
-
-   constexpr double pow10_const(int v) {
-      double r = 1.0;
-      while (v-- > 0) r *= 10.0;
-      return r;
+   inline uint64_t pow10_const(unsigned int v) {
```
Comment (elmato) [inline include/evm_runtime/types.hpp:19]:
```c++
template <std::size_t N>
constexpr std::array<intx::uint256, N> generate_N_powers_of_10() {
    std::array<intx::uint256, N> powers{};
    for (auto i=0; i < powers.size(); i++) {
        powers[i] = intx::exp(10_u256, intx::uint256(i));
    }
    return powers;
}

constexpr auto minimum_natively_representable_at = generate_N_powers_of_10<19>();

int main() {
    const auto& minimum_natively_representable = minimum_natively_representable_at[10];
    std::cout << intx::to_string(minimum_natively_representable) << std::endl;
    return 0;
}
```

Comment (elmato) [reply include/evm_runtime/types.hpp:19]: I was thinking that maybe we can generate the uint256 at compile time, and then get a reference from that table everywhere we need the min_native_representable

```
    const auto& minimum_natively_representable = minimum_natively_representable_at[10];
```

**Review by elmato — APPROVED** [review]

**Pull request merged**